### PR TITLE
Added a controllable directory.

### LINKED ISSUE #704: Use the directory abstraction to record progress and respond to kill switch
A lot of the tantivy code is currently protected by a protect guard to pass health check.
This is suboptimal as it does not make it possible to detect a dead lock that might happen in tantivy.

In addition, if the killswitch is activated, we would rather be able to stop as soon as possible these blocking functions.

To do this, we cannot stop the threads and we'd rather avoid leaking our need into tantivy.
The idea would be to wrap our Directory in order to record progress everytime we write on disk.

In addition, we could check and return an error if the kill switch has been activated.

### PR DESCRIPTION
**Pull request by fulmicoton**
The directory records progress and checks for the kill switch on every
writes.

Closes #704. See details in the issue.

### Description
Describe the proposed changes made in this PR.

### How was this PR tested?
Describe how you tested this PR.

### CODE REVIEW

**Comment by fulmicoton** [comment]
I just tested the PR and I get packager timeouts.

File: quickwit-actors/src/progress.rs [hunk 1]
```diff
@@ -85,7 +85,7 @@ impl Progress {
 
     pub fn protect_zone(&self) -> ProtectedZoneGuard {
         loop {
-            let previous_state: ProgressState = self.0.load(Ordering::SeqCst).into();
+            let previous_state: ProgressState = self.0.load(Ordering::Relaxed).into();
```
Comment (fmassot) [inline quickwit-actors/src/progress.rs:88]: Thinking about ordering is always a brainfuck. But in this simple case, I agree, I think we don't really care so `Relaxed` sounds good.

File: quickwit-indexing/failpoints/mod.rs [hunk 2]
```diff
@@ -210,3 +217,119 @@ async fn aux_test_failpoints() -> anyhow::Result<()> {
     );
     Ok(())
 }
+
+const TEST_TEXT: &'static str = r#"His sole child, my lord, and bequeathed to my
```
Comment (fmassot) [inline quickwit-indexing/failpoints/mod.rs:221]: Great text.

File: quickwit-indexing/src/actors/merge_executor.rs [hunk 3]
```diff
@@ -354,27 +377,22 @@ impl MergeExecutor {
             replaced_segments_demux_field_readers,
             demuxed_virtual_splits,
         );
-        info!("demux-build-mapping-end");
```
Comment (fmassot) [inline quickwit-indexing/src/actors/merge_executor.rs:357]: you can remove this one yes, can you also remove `info!("demux-build-mapping-start")` to be consistent, I added these when I was trying to understand which demux part was taking time.

File: quickwit-indexing/src/controllable_directory.rs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+/// with functionalities such as
+/// - records progress everytime a write (Note there is however a buffer writer above it)
+/// - if the killswitch is activated, returns an error on the first write happening after it
+/// - in the future, record a writing speed, possibly introduce some throttling, etc.
+#[derive(Clone)]
+pub struct ControllableDirectory {
+    inner: Arc<Inner>,
+}
+
+impl ControllableDirectory {
```
Comment (fmassot) [inline quickwit-indexing/src/controllable_directory.rs:47]: `ControlledDirectory` or `ControllableDirectory`?
`ControlledWrite` or `ControllableWrite` ? 
I would choose `controlled`. 
(not really important, you can ignore)

File: quickwit-indexing/src/controllable_directory.rs [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+
+    fn watch(&self, watch_callback: WatchCallback) -> tantivy::Result<WatchHandle> {
+        self.inner.controls.check_if_alive()?;
+        self.inner.underlying.watch(watch_callback)
+    }
+}
+
+impl TerminatingWrite for ControlledWrite {
+    #[inline]
+    fn terminate_ref(&mut self, token: AntiCallToken) -> io::Result<()> {
```
Comment (fmassot) [inline quickwit-indexing/src/controllable_directory.rs:202]: didn't know about this trick  `AntiCallToken` in tantivy, interesting.

File: quickwit-indexing/src/controllable_directory.rs [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+use tantivy::common::{AntiCallToken, TerminatingWrite};
+use tantivy::directory::error::{DeleteError, OpenReadError, OpenWriteError};
+use tantivy::directory::{FileHandle, WatchCallback, WatchHandle, WritePtr};
+use tantivy::Directory;
+
+/// Buffer capacity.
+///
+/// This is the current default for the BufWriter, but considering this constant
+/// will have a direct impact on health check, we'd better fix it.
+const BUFFER_NUM_BYTES: usize = 8_192;
```
Comment (fmassot) [inline quickwit-indexing/src/controllable_directory.rs:35]: `8_192` sounds like a good starting value, do you want to update the buffer size regarding the number of bytes to write?

**Review by fmassot — APPROVED** [review]
Really nice piece of code and simple.

File: quickwit-indexing/src/controllable_directory.rs (hunk 6 above)
Comment (fulmicoton) [reply quickwit-indexing/src/controllable_directory.rs:35]: No. 8KB or 16KB is considered optimal. 8K is the default in rust btw

File: quickwit-indexing/src/controllable_directory.rs (hunk 4 above)
Comment (fulmicoton) [reply quickwit-indexing/src/controllable_directory.rs:47]: Went for Controlled!

File: quickwit-indexing/src/actors/merge_executor.rs (hunk 3 above)
Comment (fulmicoton) [reply quickwit-indexing/src/actors/merge_executor.rs:357]: I did not remove it on purpose actually. Thanks for spotting!

**Pull request closed without merging**